### ISSUE TITLE
Fizsh cannot be used as login shell due to missing PATH

### ISSUE THREAD

**Issue by dbeecham** [zsh-users/fizsh#6]
When using fizsh as login shell, PATH is not set and fails:

```
[user@host ~]# ssh user@localhost
user@localhost's password:
Last login: Thu Jan  8 10:24:00 2015 from localhost.localdomain
_fizsh: Make sure that fizsh is installed correctly and that your path is set accordingly
```

I solved this by sourcing /etc/zsh/zprofile in /usr/local/bin/fizsh before $(type fizsh), but I am unsure if this is the correct solution. Should zprofile be sourced every time?

I am running current Arch Linux, zsh version 5.0.7, fizsh version 1.0.8.

**Comment by guidovansteen** [comment]
Thank you for reporting this issue. 

Currently fizsh only runs as a login shell if it is explicitly invoked as "fizsh -l" or "fizsh --login". Your ssh session calls fizsh without the l flag. This explains why the new instance of fizsh behaves as a non-login shell. This means that "zprofile" is not sourced. On Arch this means in turn that /etc/profile is not sourced either (see https://wiki.archlinux.org/index.php/zsh). The result on Arch is that no $path is set when fizsh is started, which explains the reported message. 

Most other distros do not have this problem, because their zsh does not depend on $path being set in /etc/profile. 

The issue can indeed be circumvented by manually sourcing zprofile before $(type fizsh) is evaluated.

**Comment by dbeecham** [comment]
All right, I have tried around with --login, but I cannot seem to quite figure it out. 

Firstly, I cannot set "/usr/local/bin/fizsh --login" as shell in /etc/passwd; doing to gives me an error "shell /path/fizsh --login does not exist". Not sure if I should add this to /etc/shells, of if I need to escape spaces in passwd, or if this just does not work. So, I created a script:

```
$ cat /usr/local/bin/fizsh-local
#!/usr/bin/env zsh

echo $PATH

/usr/local/bin/fizsh --login
```

I also added a `echo $PATH` at the top in fizsh.

```
/usr/bin:/bin:/usr/sbin:/sbin
```
Is the PATH echoed. So `$PATH` is not empty. But `/usr/local/bin` isn't added before `/etc/profile` is sourced. Also, `--login` in `/usr/local/bin/fizsh-local` did not work.

**Comment by dbeecham** [comment]
I am reading though the source code, trying to understand how to solve the problem. `type fizsh` is called on line 58. The only reference to --login is on line 148, but it falls into the same condition as no arguments. Am I missing something?

**Comment by guidovansteen** [comment]
Around line 170 you will find: 

`if [[ ( $@ == "" ) ]]; then 
    zsh
  else
    zsh -l
  fi`

So, if fizsh is invoked with a command-line argument (which in this case must be -l or --login) then it is started as "zsh -l" (after setting $ZDOTDIR and some other things).

**Comment by guidovansteen** [comment]
At this moment I do not have an arch machine to test things. I have one at home. I will look into it later.

**Comment by dbeecham** [comment]
Ah, I see. The shell is exiting at line 59 though. Should some code be moved around?

**Comment by guidovansteen** [comment]
> But /usr/local/bin isn't added before /etc/profile is sourced. 

So /usr/bin is in your path. In that case you could install fizsh as /usr/bin/fizsh. Try something like: 

`user@localmachine$ ./configure --bindir=/usr/bin` 

See https://github.com/zsh-users/fizsh#install

**Comment by guidovansteen** [comment]
Some further thought: 

The script /usr/bin/fizsh (in your case /usr/local/bin/fizsh) is initially invoked as a non-login shell. This makes difficult to determine whether the eventual fizsh shell should be run as a login-shell or not. Because of this fizsh requires the "-l" or "--login" option in order to be started as a login shell. 

Yet, after using "ssh -l user@...", "sudo -iu user" or "su -i user", you should get a login shell as well. As you already noted, the problem is that in this case the system logs you in through the /etc/password file, which makes it difficult to provide the -i flag. 

If we don't find a proper solution, I will add a note about this in the manpage in the "BUGS" section.

**Comment by dbeecham** [comment]
Even if fizsh is started with the `--login` argument, the shell will still fail, though? The flag is not used until the call to zsh, $(type fizsh) will be empty anyway.

I can see the issue, though. `zsh` should be doing the sourcing, but it can't do it before `fizsh` has set up all it needs to, but it can't set it up before `zsh` has done the sourcing. It's a catch 22.

I also just realized my line numbers are different from yours since I edited the file.

**Comment by dbeecham** [comment]
A thought; is it possible to use $0?

**Comment by guidovansteen** [comment]
> Even if fizsh is started with the --login argument, the shell will still fail, though? 
> The flag is not used until the call to zsh, $(type fizsh) will be empty anyway.

If fizsh is installed as `/usr/bin/fizsh` and if `/usr/bin` is part of your path `$(type fizsh)` will yield `/usr/bin/fizsh` this is fizsh it works by default on Debian. 

`/usr/bin

**Comment by guidovansteen** [comment]
My comment was not finished yet: 

If `/usr/local/bin` is added to the path at the same time as `/usr/bin` things will work as well. 

Wouldn't you encounter similar problems if zsh is only installed in `/usr/local/bin/` and `/usr/local/bin/` is not part of the path?

**Comment by guidovansteen** [comment]
> A thought; is it possible to use $0?

I do not know how to do that. I think it will be difficult. Even zsh itself does not folow the line that $0 should be "-zsh" for zsh's login-shells. 

$ `zsh -fdl` 
% `echo $0` 
zsh
% `if [[ -o login ]]; then; print yes; else; print no; fi`
yes 
% `sudo -i`
# `echo $0`
-bash

Or: 

$ `su -l user -s /usr/bin/zsh`
% `if [[ -o login ]]; then; print yes; else; print no; fi`
yes 
% `echo $0`
-su

**Comment by guidovansteen** [comment]
zsh actually distinguishes two types of login shells: (1) shells started by login and (2) shells started with the `-l` or `--login` option. 

Type 1 login shells can be detected using the environment variable `$SHLVL`. For zsh login shells `$SHLVL` is always equal to 1. This implies that for fizsh login shells `$SHLVL` is always equal to 2. The reason for this is that `/usr/bin/fizsh` is a zsh script, in which `$SHLVL` is thus 1. This script makes sure fizsh can be started with another call to `zsh`, which increases `$SHLVL` to 2.

Type 2 login shells can be detected by exporting an additional environment variable, e.g. `$_fizsh_F_LOGIN_OPTION_SET` before fizsh is started in `/usr/bin/fizsh`. After fizsh has started this exported variable could be unset, after it has been copied to a local (i.e. unexported) version of the environment variable, e.g. `$_fizsh_F_LOCAL_LOGIN_OPTION_SET`. 

Depending on the values of the environment variables `$SHLVL`, `$_fizsh_F_LOGIN_OPTION_SET`, zsh could be called with or without the `-l` flag in the script `/usr/bin/fiszh`. 

Depending on the values of the environment variables `$SHLVL`, `$_fizsh_F_LOCAL_LOGIN_OPTION_SET`, `$0` could be set to `-fizsh` or `fizsh` by `/etc/fizsh/fizshrc`. 

In the next version of fizsh I intend to implement the treatment of loginshells as outlined above.

**Comment by guidovansteen** [comment]
http://www.zsh.org/mla/workers/2007/msg00270.html and subsequent reactions provide interesting information as well. 

fizsh's man page should probably mention that things like `ssh -t someserver "cd $(pwd) && ARGV0=-\$0 exec \$0"` do not work as described there.

**Comment by guidovansteen** [comment]
Closing this issue (see the above commits).

**Comment by guidovansteen** [comment]
PS: The original issue of fizsh not being able to find its path, is IMO misconfiguration problem during the build process. 

If Pacman had done the installation, fizsh should have been installed in `/usr/bin`. This would have prevented the problem. `/usr/local/bin` is the responsability of the local administrator. If he decides to install fizsh there, he should make sure that `/usr/local/bin` is available when fizsh is started as a login shell. This may for example be achieved by including `/usr/local/bin` in the $path variable in `zprofile` or in `zlogin`.

**Comment by fibric** [comment]
On OSX the issue still persists. Due to security reasons, on OSX 10.11 /usr/bin is protected by SIP (Security Integrity Protection).

Installing 1.0.8 can be done by `brew install fizsh` and will place bin aliases into /usr/local

```
fibrics-Mac-Pro% fizsh
welcome to fizsh, the friendly interactive zshell
type man fizsh for instructions on how to use fizsh
fibric@fibrics-Mac-Pro ~>
```

But when changing my default shell from zsh to fizsh then I see this:
```
Last login: Sat Oct  3 08:32:55 on ttys001
_fizsh: Make sure that fizsh is installed correctly and that your path is set accordingly

[Process completed]
```

At the point of [Process completed] the terminal window is dead.

**Comment by guidovansteen** [comment]
Strange... This error message occurs when the command `type fizsh` cannot find the fizsh binary. Not finding the fizsh binary causes your seemingly dead terminal window. Usually this happens when the directory of the fizsh binary is not part of $path. Therefore, after you arrive at the dead terminal window, can you report the output of `type fizsh` and `echo $path`? 

I do not have `OS X` and `brew` here. So I cannot test the issue myself. 

You might want to check if the problem is related to `brew install`. To find out, just install fizsh manually: 

`mkdir ./fizsh_sources`
`cd ./fizsh_sources`
`git clone https://github.com/zsh-users/fizsh.git`
`cd fizsh`
`./configure --prefix=/usr/local --includedir=/usr/include --datadir=/usr/share --bindir=/usr/local/bin --libexecdir=/usr/lib/fizsh --localstatedir=/var --sysconfdir=/etc/local`
`make` 
`sudo make` 
`chsh $USER -s /usr/local/bin/fizsh` 
`su $USER`

If this does work, I would expect the problem to be caused by `brew install`. 

Best wishes, 

Guido

**Comment by gerstorfer** [comment]
Manual installation makes no difference.

**Comment by guidovansteen** [comment]
Could you report the output of `type fizsh` and `echo $path`?

**Comment by guidovansteen** [comment]
And could you report the ouput of the manal installation process as well?

**Comment by gerstorfer** [comment]
Manual installation reports:

```
Cloning into 'fizsh'...
remote: Counting objects: 658, done.
remote: Total 658 (delta 0), reused 0 (delta 0), pack-reused 658
Receiving objects: 100% (658/658), 334.17 KiB | 612.00 KiB/s, done.
Resolving deltas: 100% (318/318), done.
Checking connectivity... done.
checking for a BSD-compatible install... /usr/local/bin/ginstall -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/local/bin/gmkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for perl... perl
configure: creating ./config.status
config.status: creating Makefile
config.status: creating doc/Makefile
config.status: creating man/Makefile
config.status: creating scripts/Makefile
configure: creating ./config.status
config.status: creating Makefile
config.status: creating doc/Makefile
config.status: creating man/Makefile
config.status: creating scripts/Makefile
Making all in doc
make[1]: Nothing to be done for `all'.
Making all in man
pod2man --utf8 -c "" -r "fizsh 1.0.8" ./fizsh.pod > ./fizsh.1
Making all in scripts
make[1]: Nothing to be done for `all'.
make[1]: Nothing to be done for `all-am'.
Making install in doc
make[2]: Nothing to be done for `install-exec-am'.
 /usr/local/bin/gmkdir -p '/usr/local/share/doc/fizsh'
 /usr/local/bin/ginstall -c -m 644 copyright changelog '/usr/local/share/doc/fizsh'
Making install in man
make[2]: Nothing to be done for `install-exec-am'.
 /usr/local/bin/gmkdir -p '/usr/local/share/man/man1'
 /usr/local/bin/ginstall -c -m 644 fizsh.1 '/usr/local/share/man/man1'
Making install in scripts
 /usr/local/bin/gmkdir -p '/usr/local/bin'
 /usr/local/bin/ginstall -c fizsh '/usr/local/bin'
 /usr/local/bin/gmkdir -p '/etc/local/fizsh'
 /usr/local/bin/ginstall -c fizshrc.zsh zsh-history-substring-search.zsh zsh-syntax-highlighting.zsh fizsh-prompt.zsh modify-etc-shells.sh brackets-highlighter.zsh cursor-highlighter.zsh main-highlighter.zsh pattern-highlighter.zsh root-highlighter.zsh fizsh-miscellaneous.zsh '/etc/local/fizsh'
make[2]: Nothing to be done for `install-exec-am'.
./scripts/modify-etc-shells.sh --add /usr/local/bin/fizsh || true
```

`type fizsh` gives `fizsh is /usr/local/bin/fizsh`
 
`echo $path` gives `/usr/local/bin /usr/bin /bin /usr/sbin /sbin /opt/X11/bin /Library/TeX/texbin /usr/local/opt/ruby/bin /usr/local/sbin /usr/local/opt/coreutils/libexec/gnubin`

**Comment by gerstorfer** [comment]
It appears, that you are right: no $path is set. So adding `export PATH=/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin` to `/usr/local/bin/fizsh` fixes the issue.

**Comment by guidovansteen** [comment]
Thank you for reporting this. 

I have read some posts on the OSX-specific way of setting the system-wide path (https://serverfault.com/questions/16355/how-to-set-global-path-on-os-x). Directories to be included in it should be listed in the file `/etc/paths`. 

So, administrators who allow executables in `/usr/local/bin/` should add the entry `/usr/local/bin/` to the file `/etc/paths`. This should be done BEFORE any executable in `/usr/local/bin/` is called. 

Changing `PATH` in `/etc/zshenv` may technically work as well. The same may apply to changing `PATH` in `/etc/launchd.conf`. However, on OSX I would rather create the correct system-wide path using `/etc/paths`.

**Comment by guidovansteen** [comment]
It would not hurt to add an extra configuration file called `~/.fprofile`. This file would be sourced immediately after `fizsh` is started. I might add such a file in the next `fizsh` release.

**Comment by gerstorfer** [comment]
I would like to note, that I added `/usr/local/bin` to `/etc/paths` (and `/etc/launchd.conf` and `/etc/paths.d/`) without success. I tried a lot of other stuff too. In the end, using `/etc/zshenv` was the only feasible workaround.

**Comment by guidovansteen** [comment]
I have access to an OSX machine (10.8.5). On this machine there is a file called `/etc/zshenv`. It contains: 

$ `cat /etc/zshenv`
...
if [ -x /usr/libexec/path_helper ]; then
	eval \`/usr/libexec/path_helper -s\`
fi

This `path_helper` program sets the path using the contents of `/etc/paths`. In my case this seems to work properly.

**Comment by gerstorfer** [comment]
On OSX 10.11 `/etc/zshenv` does not exist and has to be created with this content.

**Comment by guidovansteen** [comment]
OK, now I understand. 

Can you set the path in `~/.zshenv`? Or is that not working either? Does the `path_helper` binary still exist in OSX 10.11?

**Comment by gerstorfer** [comment]
Yes! I can set the path in `~/.zshenv` and `path_helper` still exists and works as advertised.

**Comment by guidovansteen** [comment]
Thank you, it is good to know this! 

In order to avoid this type of trouble with `PATH`, I will add two extra configuration files to fizsh: `/etc/fizsh/fizshenv` and `~/.fizshenv`. Initially they will be like `/etc/zsh/zshenv` on Debian: 

localuser@debian > cat /etc/fizsh/fizshenv 
if [[ -z "$PATH" || "$PATH" == "/bin:/usr/bin" ]]
then
	export PATH="/usr/local/bin:/usr/bin:/bin"
fi

**Comment by guidovansteen** [comment]
Commited code that should prevent this issue (https://github.com/zsh-users/fizsh/commit/9a128f3239bd72b52e97281a3377e17b011fd8b6).

**Comment by guidovansteen** [comment]
On second thought: Since this issue has now been solved in a relatively simple way, I will refrain from adding extra configuration files like `/etc/fizsh/fizshenv` and `~/.fizsh/fizshenv`. Fizsh is meant to be a friendly shell. I will keep the number of configuration files as small as possible.

**Comment by djbclark** [comment]
I can report this does work on 10.11. Only comment is that it would be nice to have fizsh tell the user in the error message to try putting PATH in .zshenv rather than making thr user find this thread :-) (BTW great work in general, Iam really enjoying it.)
